src/stress/memory/oom.spec.ts

```diff
@@ -3,32 +3,157 @@ Stress tests covering robustness when available VRAM is exhausted.
+  )
+  .fn(async t => {
+    // Use a relatively large size to quickly hit OOM.
+    const kSize = 512 * 1024 * 1024;
+
+    const { mappable, mode, mappedAtCreation, unmapBeforeResolve } = t.params;
+    const buffers: GPUBuffer[] = [];
+    // Closure to call map and verify results on all of the buffers.
+    const finish = async () => {
+      if (mappable) {
+        await Promise.all(buffers.map(value => value.mapAsync(mode)));
+      } else {
+        buffers.forEach(value => {
+          t.expectValidationError(() => {
+            void value.mapAsync(mode);
+          });
+        });
+      }
+      // Finally, destroy all the buffers to free the resources.
+      buffers.forEach(buffer => buffer.destroy());
+    };
+
+    let errorBuffer: GPUBuffer;
+    for (;;) {
+      if (mappedAtCreation) {
+        // When mappedAtCreation is true, OOM can happen on the client which throws a RangeError. In
+        // this case, we don't do any validations on the OOM buffer.
+        try {
+          t.device.pushErrorScope('out-of-memory');
+          const buffer = createBuffer(t.device, kSize, mappable, mode, mappedAtCreation);
+          if (await t.device.popErrorScope()) {
+            errorBuffer = buffer;
+            break;
+          }
+          buffers.push(buffer);
+        } catch (ex) {
+          await finish();
+        }
+      } else {
+        t.device.pushErrorScope('out-of-memory');
+        const buffer = createBuffer(t.device, kSize, mappable, mode, mappedAtCreation);
+        if (await t.device.popErrorScope()) {
+          errorBuffer = buffer;
+          break;
+        }
+        buffers.push(buffer);
+      }
+    }
+
+    // Do some validation on the OOM buffer.
+    let promise: Promise<void>;
+    t.expectValidationError(() => {
+      promise = errorBuffer.mapAsync(mode);
+    });
+    if (unmapBeforeResolve) {
+      // Should reject with abort error because buffer will be unmapped
+      // before validation check finishes.
+      t.shouldReject('AbortError', promise!);
+    } else {
+      // Should also reject in addition to the validation error.
+      t.shouldReject('OperationError', promise!);
+
+      // Wait for validation error before unmap to ensure validation check
+      // ends before unmap.
+      try {
+        await promise!;
+        throw new Error('The promise should be rejected.');
+      } catch {
+        // Should cause an exception because the promise should be rejected.
+      }
+    }
+
+    // Should throw an OperationError because the buffer is not mapped.
+    // Note: not a RangeError because the state of the buffer is checked first.
+    t.shouldThrow('OperationError', () => {
+      errorBuffer.getMappedRange();
+    });
+
+    // Should't be a validation error even if the buffer failed to be mapped.
+    errorBuffer.unmap();
+    errorBuffer.destroy();
+
+    // Finish the rest of the test w.r.t the mappable buffers.
+    await finish();
```

Comment:
this is the second time we do finish() in the mappedAtCreation state - is that intended>

Reply:
Ah... good catch, adding a return there to skip all the other stuff.